### ISSUE TITLE
Input/output file dialog behind window

### ISSUE THREAD

**Issue by colouredmirrorball** [processing/processing#3775]
When using selectInput() or selectOutput(), the resulting file dialog window is always at the same location of the sketch' window but behind it. This is very annoying and off-putting since there is almost no visual evidence the dialog has opened, except when one thinks about looking behind the sketch' window. (3.0b5)

**Comment by benfry** [comment]
Code to reproduce? I don't even know what renderer you're using, so there's no way to track this down.

**Comment by colouredmirrorball** [comment]
Sorry. Windows 7 64 bit, Processing 3.0b5:

```
void setup()
{
  size(500,500, P2D);
  //surface.setLocation(500,500); <- causes crash (unrelated)
}

void draw()
{
}

void mousePressed()
{
  selectInput("test", "test");   //gets placed behind the sketch' window
}

void test(File input)
{
  println(input.getAbsolutePath());
}
```

**Comment by benfry** [comment]
Working fine on OS X. @JakubValtar can you see if this is Windows-specific? I suspect it's also JOGL-specific.

**Comment by JakubValtar** [comment]
Yes, the file picker opens in the background on my Win 8.1 64bit.

I also have violent flickering in the sketch window, but that's a separate issue.

**Comment by benfry** [comment]
Hm, sounds like a JOGL bug, but we own it.

**Comment by codeanticode** [comment]
Right now I don't have access to a windows machine, but I wrote a minimal test case that does not depend on Processing, available [here](https://github.com/codeanticode/jogl2-tests/blob/master/src/test/windows/DialogBehindNEWT.java), which could be used to report the issue to JOGL. I did similar thing for other upstream bugs (https://github.com/processing/processing/issues/3401, https://github.com/processing/processing/issues/3339, https://github.com/processing/processing/issues/3793)

**Comment by benfry** [comment]
Looks like we won't have an upstream fix in time for 3.0, so what should we do here? 

1. Put a message in the console saying "check behind this window for your the file dialog" 
2. Minimize the rendering window during file selection
3. Leave it as-is and continue to annoy people

**Comment by codeanticode** [comment]
I would go with 2, and then restoring the rendering window after the file selection dialog is closed.

I just opened a [bug](https://jogamp.org/bugzilla/show_bug.cgi?id=1224) in JOGL's tracker (also for the other ones), with the hope to bringing these issues to their attention. I will keep an eye on any updates on JOGL's tracker.

**Comment by benfry** [comment]
Sounds good; will you have time to implement or should I take a look?

**Comment by codeanticode** [comment]
I was looking into this... We would need to add some code inside the select*() functions that checks the renderer is PGraphicsOpenGL and the platform MACOSX, and if that is the case calls surface.setVisible(false) and surface.setVisible(true) after and before selectImpl(). I thought of using setVisible() because GLWindow does not have a function to request minimization.  It is not a particularly elegant hack, but would work until the issue is fixed in JOGL. What do you think?

**Comment by benfry** [comment]
I think it's worth trying, though I think it's Windows, not OS X, that has the problem.

**Comment by codeanticode** [comment]
ok, this https://github.com/processing/processing/commit/8c2552340f815c8cf06846b99a2c64569fa7b4a9 should do it. Don't have windows machine to test right now, if anyone can please let me know if it works.

**Comment by benfry** [comment]
That looks problematic—you're aware the `selectXxx()` methods are asynchronous, right? So that'll open the chooser and then immediately re-open the drawing area. Or was that intentional? Also better to do these once inside the two selectXxxxImpl functions, rather than in each variant that calls selectXxxxImpl.

**Comment by codeanticode** [comment]
you are right, the setVisible() calls should go inside run(). A problem is that we need to look at g to determine if the renderer is PGraphicsOpenGL, but selectXxxxImpl() are static.

**Comment by codeanticode** [comment]
Ok, I added a PApplet argument to selectXxxx() with commit https://github.com/processing/processing/commit/202bae8c9134299042dab0f3f82ad66378e319d6, so the check can be done in the thread that launches the dialog and collects the user selection. I noticed these methods are also used in Archiver.java, Welcome.java, and the PreferencesFrame.java.

I run some tests with platform == MACOSX, and the rendering window disappears/appears while selecting the file, and the other functionality in the PDE (preferences, sketchbook folder selection) doesn't seem to be affected. Committed changes use platform == WINDOWS. 

Take a look and let me know if this could work as a temp workaround until JOGL fixes the upstream bug.
